### DISCUSSION TITLE
High Availability and Distributed FAISS

### DISCUSSION THREAD

**Discussion by ashuezy** [facebookresearch/faiss#1677] [General]
I am aware that FAISS is a file system database and returns an id for every record inserted. My question is related to the best practice/strategy for implementing HA and make it distributed. 

My strategy:
1. I am using FAISS on an HTTP server with 5 nodes i.e A,B,C,D,E [Each machine has an HTTP server]
2. For handling insertion only **node A** is used and other nodes are for matching only.
All insertion requests are duplicated on other nodes.
In case **node A** is down. Next node is selected as master i.e **node B** and insertion requests are duplicated on other nodes.
3. When **node A** is back online at startup it finds who is the current **master node**, all insertion is blocked on all nodes till files are copied from **Node B** to **Node A**. Once files are sync by comparing MD5 servers are back to handling insertion at master node i.e **node B**.

This is all in theory and before writing code to handle this scenario I am thinking can all this be made simpler?
Is this too complex? Can this be made better?

Thanks.

**Comment by mdouze** [comment]
I am not sure we are talking about the same software package. 
Faiss is not a file system database and it does not handle records. 
Thus you cannot compute MD5 on files unless the Faiss index is dumped to disk.

**Comment by ashuezy** [comment]
Yes, I am writing indexes on disk.
I just want your opinion regarding the strategy. I plan to open-source this application later on.

**Comment by ashuezy** [comment]
Gentle reminder

**Comment by ashuezy** [comment]
Gentle Reminder#2

**Comment by ashuezy** [comment]
I have implemented a simple replication technique in order to implement HA/Distributed Architecture for matching vectors.
But I am having a problem with adding vectors as adding a vector should generate a unique key across all machines but since i am replicating this is increasing latency. So using a master machine where vector can be added and slave machines where only searching is happening. There has to be a better strategy to handle this problem.

**Comment by ashuezy** [comment]
I would love to hear some of your thoughts.

**Comment by wickedfoo** [comment]
Implementing a distributed layer on top of Faiss is not something that we, the Faiss library authors, can help you with, sorry. You might want to look into already existing third party libraries which may do something similar, such as Milvus.

**Comment by dexception** [comment]
The question is about strategy and not asking you to implement. What do you think would be the best strategy since you guys know about faiss from deep down?

**Comment by mausch** [comment]
[Milvus](https://milvus.io/) and [OpenSearch](https://opensearch.org/docs/latest/search-plugins/knn/approximate-knn/) are some examples of distributed Faiss.

[Milvus even provides a tool for migrating from Faiss](https://milvus.io/docs/f2m.md).
